File: leetcode/lc-us/cpp/1630.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    vector<bool> checkArithmeticSubarrays(vector<int> &nums, vector<int> &l, vector<int> &r) {
        int m = (int) l.size();
        vector<bool> res(m);
        for (int i = 0; i < m; i++) {
            int left = l[i], right = r[i];
            int mn = 1e5, mx = -1e5;
            for (int j = left; j <= right; j++) {
                mn = min(mn, nums[j]);
                mx = max(mx, nums[j]);
            }
            if ((mx - mn) % (right - left)) res[i] = false;
            else if (mx == mn) res[i] = true;
            else {
                int diff = (mx - mn) / (right - left);
                bool st[mx - mn + 1];
                memset(st, 0, sizeof st);
                for (int j = left; j <= right; j++) {
                    st[nums[j] - mn] = true;
                }
                bool flag = true;
                for (int x = mn + diff; x < mx; x += diff) {
                    if (!st[x - mn]) {
                        flag = false;
                        break;
                    }
                }
                res[i] = flag;
            }
        }
        return res;
    }
};

File: leetcode/lc-us/rust/1630.rs
use std::cmp::{max, min};

impl Solution {
    pub fn check_arithmetic_subarrays(nums: Vec<i32>, l: Vec<i32>, r: Vec<i32>) -> Vec<bool> {
        let m = l.len();
        let mut res = vec![false; m];
        for i in 0..m {
            let left = l[i];
            let right = r[i];
            let mut mn = 1e5 as i32;
            let mut mx = -1e5 as i32;
            for j in left..right + 1 {
                mn = min(mn, nums[j as usize]);
                mx = max(mx, nums[j as usize]);
            }
            if (mx - mn) % (right - left) != 0 {
                continue;
            } else if mn == mx {
                res[i] = true;
            } else {
                let diff = (mx - mn) / (right - left);
                let mut st = vec![false; (mx - mn + 1) as usize];
                for j in left..right + 1 {
                    st[(nums[j as usize] - mn) as usize] = true;
                }
                let mut flag = true;
                let mut x = mn + diff;
                while x < mx {
                    if !st[(x - mn) as usize] {
                        flag = false;
                        break;
                    }
                    x += diff;
                }
                res[i] = flag;
            }
        }
        res
    }
}